packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -740,10 +750,109 @@ export function runConnectedCallback(vm: VM) {
     }
 }
 
+function setupContext(vm: VM) {
+    const contextKeys = getContextKeys();
+
+    if (!contextKeys) {
+        return;
+    }
+
+    const { connectContext, contextEventKey } = contextKeys;
+    const { component } = vm;
+    const enumerableKeys = keys(getPrototypeOf(component));
+    const contextfulFieldsOrProps = enumerableKeys.filter(
+        (propName) => (component as any)[propName]?.[connectContext]
+    );
+
+    if (contextfulFieldsOrProps.length === 0) {
+        return;
+    }
+
+    let isProvidingContext = false;
+    const providedContextVarieties = new Map<unknown, Signal<unknown>>();
+    const contextRuntimeAdapter = {
+        isServerSide: false,
+        component,
+        provideContext<T extends object>(
+            contextVariety: T,
+            providedContextSignal: Signal<unknown>
+        ): void {
+            if (!isProvidingContext) {
+                isProvidingContext = true;
+
+                component.addEventListener(ContextEventName, (event: any) => {
```

Comment:
We typically use the `renderer` API for this. Although it's unclear to me how this would work in an SSR context since we don't implement an event listener shim. For the `ContextProvider` API we implemented a full SSR-compatible shim to make it work (https://github.com/salesforce/lwc/pull/3165); presumably we ought to do the same here if this needs to run server-side.